typescript/core/src/parse.ts

```diff
@@ -1,374 +1,415 @@
-    }
+  reader.offset = startOffset + recordLength;
 
-    case Opcode.FOOTER: {
-      const summaryStart = reader.uint64();
-      const summaryOffsetStart = reader.uint64();
-      const summaryCrc = reader.uint32();
-      const record: TypedMcapRecord = {
-        type: "Footer",
-        summaryStart,
-        summaryOffsetStart,
-        summaryCrc,
-      };
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+  return {
+    type: "Channel",
+    id: channelId,
+    schemaId,
+    topic: topicName,
+    messageEncoding,
+    metadata,
+  };
+}
 
-    case Opcode.SCHEMA: {
-      const id = reader.uint16();
-      const name = reader.string();
-      const encoding = reader.string();
-      const dataLen = reader.uint32();
-      if (reader.offset + dataLen > recordView.byteLength) {
-        throw new Error(`Schema data length ${dataLen} exceeds bounds of record`);
-      }
-      const data = new Uint8Array(
-        recordView.buffer,
-        recordView.byteOffset + reader.offset,
-        dataLen,
-      ).slice();
-      reader.offset += dataLen;
-
-      const record: TypedMcapRecord = {
-        type: "Schema",
-        id,
-        encoding,
-        name,
-        data,
-      };
-
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+function parseMessage(reader: Reader, recordLength: number): TypedMcapRecord {
+  const channelId = reader.uint16();
+  const sequence = reader.uint32();
+  const logTime = reader.uint64();
+  const publishTime = reader.uint64();
+  const data = reader.u8ArrayCopy(recordLength - 22 /*channelId, sequence, logTime, publishTime*/);
+  return {
+    type: "Message",
+    channelId,
+    sequence,
+    logTime,
+    publishTime,
+    data,
+  };
+}
 
-    case Opcode.CHANNEL: {
-      const channelId = reader.uint16();
-      const schemaId = reader.uint16();
-      const topicName = reader.string();
-      const messageEncoding = reader.string();
-      const metadata = reader.map(
-        (r) => r.string(),
-        (r) => r.string(),
-      );
-
-      const record: TypedMcapRecord = {
-        type: "Channel",
-        id: channelId,
-        schemaId,
-        topic: topicName,
-        messageEncoding,
-        metadata,
-      };
-
-      return { record, usedBytes: recordEndOffset - startOffset };
-    }
+function parseChunk(reader: Reader, recordLength: number): TypedMcapRecord {
+  const start = reader.offset;
+  const startTime = reader.uint64();
+  const endTime = reader.uint64();
+  const uncompressedSize = reader.uint64();
+  const uncompressedCrc = reader.uint32();
+  const compression = reader.string();
+  const recordByteLength = Number(reader.uint64());
+  const end = reader.offset;
+  if (recordLength - (end - start) < recordByteLength) {
+    throw new Error("Chunk records length exceeds remaining record size");
+  }
+  const records = reader.u8ArrayCopy(recordByteLength);
```

Comment:
I wonder if we should actually defer to the u8ArrayCopy method for bounds-checking? Now that the Reader class is being used for this, it feels redundant for the bounds check to be done by the caller

Reply:
Because throwable functions are harder for v8 to optimize / not-cost-free, I plan to shift error handling up not down the call tree.

In the optimistic/happy/common paths we should never throw here, so IMO it can be handled ad-hoc, e.g: return `null` then run ad-hoc code to determine precisely what the error was.

The indirection might be a little counterintuitive, but will help keep hot path call tree fast / simple